### ISSUE TITLE
请问有没有可能基于UniXcoder训练支持中文的模型呢

### ISSUE THREAD

**Issue by edisonh** [microsoft/CodeBERT#159]
可以通过进一步mlm做到吗

**Comment by guoday** [comment]
可能需要有中文 NL的code-nl pair在原来的代码上，对现有的unixcoder继续预训练

**Comment by edisonh** [comment]
“原来的代码”是指unixcoder原来预训练时使用的训练数据吗？

**Comment by guoday** [comment]
不是，就是预训练代码，然后再中文的code-nl pair预训练数据上训练

**Comment by edisonh** [comment]
在预训练模型上用中文的code-nl pair进行mlm，这样可以吗。因为你们好像没有公开预训练代码吧

**Comment by guoday** [comment]
可以使用MLM

**Comment by edisonh** [comment]
@guoday 感谢解答

**Comment by edisonh** [comment]
@guoday 求助大佬，能不能给一段简单的示例代码

**Comment by edisonh** [comment]
@guoday 自有数据已按code_tokens 和docstring_tokens预处理好，code_tokens使用csnet相同的处理方法，docstring_tokens使用jieba分词。看了下huggingface的适用roberta的lm_run_mlm.py，数据结构不同，不能直接使用的。

**Comment by guoday** [comment]
因为目前还无法开源预训练代码，只能给你一个MLM预训练的类函数[model.py.zip](https://github.com/microsoft/CodeBERT/files/9486626/model.py.zip)。其中类函数内的encoder为RobertaModel，数据处理成`<s> <encoder-only> </s> code_tokens </s> docstring_tokens </s>`的格式并转成id即可。

**Comment by edisonh** [comment]
@guoday 感谢大佬

**Comment by edisonh** [comment]
@guoday 大佬，作为encoder-decoder mode，论文中使用的是DNS进行预训练，下游生成类任务使用encoder-decoder mode。现在我们使用中文code-nl的进行MLM预训练encoder，能帮助下游任务生成类任务在nl为中文的数据上受益吗

**Comment by edisonh** [comment]
@guoday 大佬，还有一个问题，UniXcoder预训练是按C4、CSN unimodal code、CSN multi-modal data顺序进行训练，相同数据不按这个顺序，对结果影响大吗

**Comment by guoday** [comment]
1. 你训练完MLM后可以对比一下，这些不太清楚
2. 之所以按这个顺序进行训练是因为C4和CSN unimodal code是单语料的，先利用这一部分的数据，之后再用CSN multi-modal data双语的方式预训练。如果是只有一批数据，直接用就可以了

**Comment by edisonh** [comment]
@guoday  感谢大佬耐心解答

**Comment by edisonh** [comment]
@guoday 大佬，硬件太差了，训练了很久还没结束。。。 又看下下大佬前面给的mlm源码，有一行代码没太理解
`
...
    probability_matrix.masked_fill_(torch.tensor(special_tokens_mask, dtype=torch.bool).to(inputs.device), value=0.0)
    padding_mask = labels.lt(119)   # 这行和下一行作用什么，是消除标点符号吗？
    probability_matrix.masked_fill_(padding_mask, value=0.0)  
...
`

**Comment by edisonh** [comment]
> @guoday 大佬，硬件太差了，训练了很久还没结束。。。 又看下下大佬前面给的mlm源码，有一行代码没太理解
> 
> ```
> ...
>     probability_matrix.masked_fill_(torch.tensor(special_tokens_mask, dtype=torch.bool).to(inputs.device), value=0.0)
>     padding_mask = labels.lt(119)   # 这行和下一行作用什么，是消除标点符号吗？
>     probability_matrix.masked_fill_(padding_mask, value=0.0)  
> ...
> ```

@guoday  debug了下，了解了，是盖住<encoder-only>这种预留的mask

**Comment by SmartMapple** [comment]
> > @guoday 大佬，硬件太差了，训练了很久还没结束。。。 又看下下大佬前面给的mlm源码，有一行代码没太理解
> > ```
> > ...
> >     probability_matrix.masked_fill_(torch.tensor(special_tokens_mask, dtype=torch.bool).to(inputs.device), value=0.0)
> >     padding_mask = labels.lt(119)   # 这行和下一行作用什么，是消除标点符号吗？
> >     probability_matrix.masked_fill_(padding_mask, value=0.0)  
> > ...
> > ```
> 
> @guoday debug了下，了解了，是盖住这种预留的mask

这个model.py是transformers里的源码

**Comment by guoday** [comment]
> > @guoday 大佬，硬件太差了，训练了很久还没结束。。。 又看下下大佬前面给的mlm源码，有一行代码没太理解
> > ```
> > ...
> >     probability_matrix.masked_fill_(torch.tensor(special_tokens_mask, dtype=torch.bool).to(inputs.device), value=0.0)
> >     padding_mask = labels.lt(119)   # 这行和下一行作用什么，是消除标点符号吗？
> >     probability_matrix.masked_fill_(padding_mask, value=0.0)  
> > ...
> > ```
> 
> @guoday debug了下，了解了，是盖住这种预留的mask

是的，因为119以下的都是特殊符号

**Comment by edisonh** [comment]
> > > @guoday 大佬，硬件太差了，训练了很久还没结束。。。 又看下下大佬前面给的mlm源码，有一行代码没太理解
> > > ```
> > > ...
> > >     probability_matrix.masked_fill_(torch.tensor(special_tokens_mask, dtype=torch.bool).to(inputs.device), value=0.0)
> > >     padding_mask = labels.lt(119)   # 这行和下一行作用什么，是消除标点符号吗？
> > >     probability_matrix.masked_fill_(padding_mask, value=0.0)  
> > > ...
> > > ```
> > 
> > 
> > @guoday debug了下，了解了，是盖住这种预留的mask
> 
> 是的，因为119以下的都是特殊符号

@guoday 感谢大佬回复

**Comment by edisonh** [comment]
> > > @guoday 大佬，硬件太差了，训练了很久还没结束。。。 又看下下大佬前面给的mlm源码，有一行代码没太理解
> > > ```
> > > ...
> > >     probability_matrix.masked_fill_(torch.tensor(special_tokens_mask, dtype=torch.bool).to(inputs.device), value=0.0)
> > >     padding_mask = labels.lt(119)   # 这行和下一行作用什么，是消除标点符号吗？
> > >     probability_matrix.masked_fill_(padding_mask, value=0.0)  
> > > ...
> > > ```
> > 
> > 
> > @guoday debug了下，了解了，是盖住这种预留的mask
> 
> 这个model.py是transformers里的源码

mask_tokens 基本是来自DataCollatorForLanguageModeling的，但是119这段是根据unixcoder自己的mask code增加的。

model部分则和RobertaForMaskedLM 不一样了
